lib/Dialect/Comb/CombFolds.cpp

```diff
@@ -758,6 +758,46 @@ OpFoldResult AndOp::fold(ArrayRef<Attribute> constants) {
   return constFoldAssociativeOp(constants, hw::PEO::And);
 }
 
+/// returns a single common operand that all inputs of the operation `op` can
+/// be traced back to, or an empty `Value` if no such operand exists.
+///
+/// for example for `or(a[0], a[1], ..., a[n-1])` this function returns `a`
+/// (assuming the bit-width of `a` is `n`)
+template <typename Op>
+static Value getCommonOperand(Op op) {
+  if (!op.getType().isInteger(1))
+    return Value();
+
+  auto inputs = op.inputs();
+  size_t size = inputs.size();
+
+  auto sourceOp = inputs[0].template getDefiningOp<ExtractOp>();
+  if (!sourceOp)
+    return Value();
+  Value source = sourceOp.getOperand();
+
+  // fast path: the source bit-width does not match the size of the inputs
+  if (!source || size != source.getType().getIntOrFloatBitWidth())
+    return Value();
+
+  // tracks the bits that were encountered
+  auto bits = APInt::getZero(size);
+  bits.setBit(sourceOp.lowBit());
+
+  for (size_t i = 1; i != size; ++i) {
+    if (auto extractOp = inputs[i].template getDefiningOp<ExtractOp>()) {
+      if (extractOp.getOperand() != source)
+        return Value();
+
+      bits.setBit(extractOp.lowBit());
+    } else {
+      return Value();
+    }
```

Comment:
nit: control flow is a bit complex. Could you change like this?
```suggestion
   auto extractOp = inputs[i].template getDefiningOp<ExtractOp>();
    if (!extactOp || extractOp.getOperand() != source)
      return Value();
   bits.setBit(extractOp.lowBit());
```